feat: static importing of controllers

ankush opened the pull request:
This PR adds a way to directly initialize a document by using the controller class. Makes some progress towards https://github.com/frappe/frappe/issues/39482.

This won't be ported to v16 as we'd eventually want subtle breaking changes. 



| Old way                                            | New way                                    | Comment                                   |
| ---                                                | ---                                        | ---                                       |
| `frappe.get_doc("ToDo", "T-001")`                  | `ToDo.docs.get("T-001")`                   | Get doc by name                           |
| `frappe.get_last_doc("ToDo")`                      | `ToDo.docs.last()`                         | Get last doc                              |
| `frappe.get_doc("ToDo", {"reference_doc": "ABC"})` | `ToDo.docs.last({"reference_doc": "ABC"})` | Get doc by filter (explicit)              |
| `frappe.get_docs("ToDo", {"status": "Open"})`      | `ToDo.docs.filter({"status": "Open"})`     | Get multiple docs efficiently.            |
| `frappe.new_doc("ToDo", {"status": "Open"})`       | `ToDo.docs.new(status="Open")`             | Create a document explicitly              |
| `frappe.get_cached_doc("ToDo", "T-001")`           | `ToDo.docs.get("T-001", cached=True)`      | Get redis cached version                  |
| `frappe.get_lazy_doc("ToDo", "T-001")`             | `ToDo.docs.get("T-001", lazy=True)`        | Get lazy document                         |
| `frappe.delete_doc("ToDo", "T-001")`               | `ToDo.get("T-001").delete()`                | delete document                           |
| `frappe.rename_doc("ToDo", "T-001", "T-002")`      | `ToDo.docs.get("T-001").rename("T-002")`   | rename document                           |



See commits that migrate `desk` module to evaluate how this change "feels".

Alchez commented:
Remembering import paths for doctypes could be a significant trade-off, I would think, unless it was coupled with better namespacing for the doctype controllers, which would require a lot of effort and restructuring.

If we still wanted to use the ORM and not bother with the imports, could we consider creating a simple dynamic map object between doctype strings and class controllers? When we use `get_doc`, the system already figures out which doctype controller to return. If we just store these somewhere accessible, then imports can stay as it is, and we get the benefit from Python analyzers too (although we'd lose the performance benefit, but that might be minimal anyway).

I'm not sure how much of this is actually doable, but just my $0.02.

ankush commented:
About remembering import paths.

1. IDEs/Editors with LSP can do imports automatically.  YMMV so If nothing works, configure snippets. 😄  E.g. I use pyright, pycharm also has similar feature.
![Screenshot 2021-05-21 at 7 17 02 PM](https://user-images.githubusercontent.com/9079960/119147358-229d5200-ba69-11eb-9947-50621b6ed047.png)

2. It's a one-time job vs  time spent on looking up method/variable name on the dynamically imported object. 
3. Some of the modules are already being imported for utility functions defined in controllers.

I don't think moving away from one dynamic solution to another would help us much.

chromonav commented:
I am in favor of static imports. But would this mean `get_doc` etc. would be deprecated?

ankush commented:
@chromonav no, it will always be there. Just won't be recommend unless you need dynamic behaviour i.e. cases where you don't even know which doctype you're operating on.

ankush committed:
feat: static importing of DocType

Initilaize a controller class directly if class has _DOCTYPE_NAME
present.

ankush committed:
feat: new doc using dict from controller class

Directly create new doc using dict as argument to controller class
initializer.

No need to add `doctype` key when using class for initalizing.

sqash this

ankush committed:
feat: Controller.new() classmethod for new doc

Create new doc using controller class. Return value is annotated to be
of same type as class.

ankush committed:
feat: cached doc using Controller.from_cache(name)

ankush committed:
refactor: use generic type instead of cls

ankush committed:
refactor: eliminate `return` from `__init__`

It is not recommended to have `return` in __init__ methods, while the
refactor looks complicated, it still accurately captures the behaviour.

ankush committed:
refactor: reduce use of hasattr

ankush committed:
feat: directly initialize single controllers

e.g.

from erpnext.stock.doctype.stock_settings.stock_settings import StockSettings

setting = StockSettings()

ankush commented:
One way to shorten the long import path is by adding module doctypes in `module/doctype/__init__.py` file 🤔 

E.g. 

```
# file erpnext/stock/doctype/__init__.py

from .item.item import Item
```

This allows shortening the import path

`from erpnext.stock.doctype import Item`


I'll evaluate this further. This will probably cause issues because of clashes in symbol names. Also, this will encourage use of `import *` 😆

ankush committed:
fix: throw error if doctype key is not specified

vrslev commented:
@ankush What about keeping `get_doc` and `new_doc` for default but return document type explicity?

rmehta commented:
Convert a few files where this is directly applied so that the tests can verify the code

gavindsouza commented:
How are extending standard controllers handled?

ref: https://frappeframework.com/docs/user/en/python-api/hooks#override-doctype-class

ankush commented:
@gavindsouza  haven't considered it so far. But it seems to work like normal inheritance works in python with the exception being parent classes will still get all the fields (including new fields added by a child class, this is because fields are not defined in python). 

To use extended methods, the extended class has to be used while creating objects.


Example:
![Screenshot 2021-07-02 at 3 09 49 PM](https://user-images.githubusercontent.com/9079960/124255319-a73ebe00-db47-11eb-8352-62d2ed5a578d.png)


`get_doc` behaviour remains unchanged:

![get doc override](https://user-images.githubusercontent.com/9079960/124260718-a1e47200-db4d-11eb-81be-8a7d593dd9f9.jpg)

rmehta commented:
@ankush As discussed, let's covert all possible imports to static imports in frappe/frappe. So that it becomes the default option.

ankush commented:
To simplify reviewing, I created separate PR for refactors by basing on this branch.  ^

rmehta commented:
@ankush whats happening with this?

ankush committed:
fix: use new_doc instead of new to avoid naming issues

ankush committed:
fix: support instantiating overriden class

ankush commented:
No bandwidth rn. reopened just to confirm the fix for overriding classes passes :p

rmehta commented:
Still a good project to pursue ;-)

ankush commented:
Yeah, I'll add this. It's not a major change. I had a chat with @netchampfaris today.

These set of commits are unworkable though. I'll reset ig.

ankush commented:
:new_moon_with_face:

ankush committed:
refactor: Specify doctype name on controllers everywhere

Only way we can find out which real `DocType` this class belongs to.

ankush committed:
feat: New `Controller.docs` collections APIs

ankush committed:
perf: don't resolve doctype name many times

ankush committed:
test: avoid manual cleanups

ankush committed:
fix: Force `name` to be int/string/None

ankush committed:
refactor: more code cleanups

- Use self._doctype directly.
- remove delete API, not required for collections... yet.

ankush committed:
test: Use new APIs for testing new features

DUH

ankush committed:
perf: add slots on DocsCollection

ankush committed:
fix: Restrict access to DocsCollection on instances

ankush committed:
fix: Skip docs descriptor

For now

ankush commented:
TODO: 

- [x] Codegen (command + dynamic updates) for migrating existing doctypes and adding `_DOCTYPE_NAME`
- [ ] Codegen (command + dynamic updates) for updating `module/doctype/__init__.py` and `module/doctype/__init__.pyi`
- [ ] Reusable "prompt" (?) for migrating modules.
